### ISSUE TITLE
panic: runtime error: invalid memory address or nil pointer dereference (maybe when another cluster node is removed)

### ISSUE THREAD

**Issue by jpetazzo** [docker/libnetwork#985]
I just experienced the following crash of my engine.

It looks like this happens has I forcibly removed another node from the cluster (essentially I powercycled `node3` and noticed that I had lost my Swarm leader, even though `node3` was *not* leader at that time; then I noticed that I had lost `node2`: the Engine had crashed).

```
Mar 02 20:44:17 node2 docker[2728]: panic: runtime error: invalid memory address or nil pointer dereference
Mar 02 20:44:17 node2 docker[2728]: [signal 0xb code=0x1 addr=0x0 pc=0xb6c858]
Mar 02 20:44:17 node2 docker[2728]: goroutine 976 [running]:
Mar 02 20:44:17 node2 docker[2728]: github.com/docker/libnetwork/drivers/overlay.(*network).endpoint(0x0, 0xc8202fa1c0, 0x40, 0x0)
Mar 02 20:44:17 node2 docker[2728]: /usr/src/docker/vendor/src/github.com/docker/libnetwork/drivers/overlay/ov_endpoint.go:23 +0xe8
Mar 02 20:44:17 node2 docker[2728]: github.com/docker/libnetwork/drivers/overlay.(*driver).notifyEvent(0xc820081860, 0x1655770, 0x4, 0xc8202fa1c0, 0x40, 0xc821464540, 0x40)
Mar 02 20:44:17 node2 docker[2728]: /usr/src/docker/vendor/src/github.com/docker/libnetwork/drivers/overlay/ov_serf.go:85 +0x77
Mar 02 20:44:17 node2 docker[2728]: github.com/docker/libnetwork/drivers/overlay.(*driver).startSerfLoop(0xc820081860, 0xc8211ad800, 0xc8211adc80, 0xc8211adce0)
Mar 02 20:44:17 node2 docker[2728]: /usr/src/docker/vendor/src/github.com/docker/libnetwork/drivers/overlay/ov_serf.go:194 +0x196
Mar 02 20:44:17 node2 docker[2728]: created by github.com/docker/libnetwork/drivers/overlay.(*driver).serfInit
Mar 02 20:44:17 node2 docker[2728]: /usr/src/docker/vendor/src/github.com/docker/libnetwork/drivers/overlay/ov_serf.go:68 +0x496
```

Full log is attached. I'll add details if it happens again (I'm doing a lot of chaos-monkey style failures so maybe it'll happen again...)

```
Server:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   c3959b1
 Built:        Mon Feb 22 21:40:35 2016
 OS/Arch:      linux/amd64
```
[docker.log.txt](https://github.com/docker/libnetwork/files/155643/docker.log.txt)

**Comment by pdevine** [comment]
I hit something really similar while attempting to set up host discovery with UCP/Swarm.  I was doing some crazy stuff to try and diagnose a problem with host discovery, so I'm not exactly sure what triggered it.  It could be that I updated the clock, upgraded the engine or even messed around with the TLS certs.


> panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x38 pc=0xb71a41]

goroutine 1548 [running]:
github.com/docker/libnetwork/drivers/overlay.(*driver).notifyEvent(0xc82068bf40, 0x1648d30, 0x4, 0xc8214fcd00, 0x40, 0xc82107a740, 0x40)
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/drivers/overlay/ov_serf.go:87 +0x81
github.com/docker/libnetwork/drivers/overlay.(*driver).startSerfLoop(0xc82068bf40, 0xc82084d740, 0xc820bd0180, 0xc820bd01e0)
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/drivers/overlay/ov_serf.go:194 +0x196
created by github.com/docker/libnetwork/drivers/overlay.(*driver).serfInit
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/drivers/overlay/ov_serf.go:68 +0x496

goroutine 1 [chan receive, 81 minutes]:
main.(*DaemonCli).CmdDaemon(0xc82030fa40, 0xc82000a090, 0x0, 0x0, 0x0, 0x0)
        /usr/src/docker/docker/daemon.go:305 +0x20a7
reflect.callMethod(0xc820644de0, 0xc820705c78)
        /usr/local/go/src/reflect/value.go:628 +0x1fc
reflect.methodValueCall(0xc82000a090, 0x0, 0x0, 0x1, 0xc820644de0, 0x0, 0x0, 0xc820644de0, 0xc820646990, 0x47a034, ...)
        /usr/local/go/src/reflect/asm_amd64.s:29 +0x36
github.com/docker/docker/cli.(*Cli).Run(0xc820644d50, 0xc82000a090, 0x1, 0x1, 0x0, 0x0)
        /usr/src/docker/.gopath/src/github.com/docker/docker/cli/cli.go:89 +0x383
main.main()
        /usr/src/docker/docker/docker.go:63 +0x43c

goroutine 17 [syscall, 84 minutes, locked to thread]:
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:1721 +0x1

goroutine 12 [chan receive, 84 minutes]:
github.com/docker/docker/daemon.setupDumpStackTrap.func1(0xc820146cc0)
        /usr/src/docker/.gopath/src/github.com/docker/docker/daemon/debugtrap_unix.go:17 +0x4d
created by github.com/docker/docker/daemon.setupDumpStackTrap
        /usr/src/docker/.gopath/src/github.com/docker/docker/daemon/debugtrap_unix.go:20 +0x12a

goroutine 6 [syscall, 84 minutes]:
os/signal.loop()
        /usr/local/go/src/os/signal/signal_unix.go:22 +0x18
created by os/signal.init.1
        /usr/local/go/src/os/signal/signal_unix.go:28 +0x37
goroutine 11 [select, 81 minutes, locked to thread]:
runtime.gopark(0x194c220, 0xc820019728, 0x164bbf0, 0x6, 0x18, 0x2)
        /usr/local/go/src/runtime/proc.go:185 +0x163
runtime.selectgoImpl(0xc820019728, 0x0, 0x18)
        /usr/local/go/src/runtime/select.go:392 +0xa64
runtime.selectgo(0xc820019728)
        /usr/local/go/src/runtime/select.go:212 +0x12
runtime.ensureSigM.func1()
        /usr/local/go/src/runtime/signal1_unix.go:227 +0x353
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:1721 +0x1

goroutine 24 [select, 9 minutes]:
github.com/docker/libnetwork.(*controller).watchLoop(0xc8208a54a0)
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/store.go:409 +0x140
created by github.com/docker/libnetwork.(*controller).startWatch
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/store.go:423 +0x100

goroutine 22 [select]:
github.com/docker/docker/daemon.(*daemonDiscoveryReloader).advertiseHeartbeat(0xc8201e6fa0, 0xc8202dc360, 0x14)
        /usr/src/docker/.gopath/src/github.com/docker/docker/daemon/discovery.go:113 +0x10f
created by github.com/docker/docker/daemon.(*daemonDiscoveryReloader).advertise
        /usr/src/docker/.gopath/src/github.com/docker/docker/daemon/discovery.go:99 +0x6b

goroutine 39 [chan receive, 84 minutes]:
github.com/docker/libnetwork/iptables.signalHandler()
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/iptables/firewalld.go:92 +0x54
created by github.com/docker/libnetwork/iptables.FirewalldInit
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/iptables/firewalld.go:48 +0x17b

goroutine 25 [select]:
github.com/docker/docker/pkg/discovery/kv.(*Discovery).watchOnce(0xc8201fdc70, 0xc8208462a0, 0xc820b823c0, 0xc820846360, 0xc820846420, 0x0)
        /usr/src/docker/.gopath/src/github.com/docker/docker/pkg/discovery/kv/kv.go:109 +0x5a9
github.com/docker/docker/pkg/discovery/kv.(*Discovery).Watch.func1(0xc820846360, 0xc820846420, 0xc8201fdc70, 0xc8208462a0)
        /usr/src/docker/.gopath/src/github.com/docker/docker/pkg/discovery/kv/kv.go:164 +0x3e7
created by github.com/docker/docker/pkg/discovery/kv.(*Discovery).Watch
        /usr/src/docker/.gopath/src/github.com/docker/docker/pkg/discovery/kv/kv.go:174 +0x98
goroutine 26 [chan send]:
github.com/docker/libnetwork/drivers/overlay.(*driver).pushLocalEndpointEvent(0xc82068bf40, 0x1648d30, 0x4, 0xc82107a740, 0x40, 0xc820b31bc0, 0x40)
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/drivers/overlay/overlay.go:191 +0xbc
github.com/docker/libnetwork/drivers/overlay.(*driver).pushLocalDb.func1(0xc82107a740, 0x40, 0xc8219112c0, 0xc8211ed450, 0xc8218e61c0)
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/drivers/overlay/peerdb.go:325 +0x6d
github.com/docker/libnetwork/drivers/overlay.(*driver).peerDbWalk.func1(0xc8219112c0, 0xc8211ed450, 0xc820bb9608)
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/drivers/overlay/peerdb.go:70 +0x48
github.com/docker/libnetwork/drivers/overlay.(*driver).peerDbNetworkWalk(0xc82068bf40, 0xc82107a740, 0x40, 0xc820bb9790, 0x0, 0x0)
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/drivers/overlay/peerdb.go:92 +0x3ee
github.com/docker/libnetwork/drivers/overlay.(*driver).peerDbWalk(0xc82068bf40, 0xc820bb9848, 0x0, 0x0)
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/drivers/overlay/peerdb.go:71 +0x270
github.com/docker/libnetwork/drivers/overlay.(*driver).pushLocalDb(0xc82068bf40)
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/drivers/overlay/peerdb.go:328 +0x39
github.com/docker/libnetwork/drivers/overlay.(*driver).nodeJoin.func1()
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/drivers/overlay/overlay.go:170 +0x96
sync.(*Once).Do(0xc82068bfcc, 0xc820bb99c8)
        /usr/local/go/src/sync/once.go:44 +0xe4
github.com/docker/libnetwork/drivers/overlay.(*driver).nodeJoin(0xc82068bf40, 0xc82160d1e0, 0xe, 0xc8211b3a01)
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/drivers/overlay/overlay.go:172 +0x42f
github.com/docker/libnetwork/drivers/overlay.(*driver).DiscoverNew(0xc82068bf40, 0x1, 0x143e500, 0xc82112f500, 0x0, 0x0)
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/drivers/overlay/overlay.go:201 +0xdf
github.com/docker/libnetwork.(*controller).pushNodeDiscovery(0xc8208a54a0, 0xc820927ba0, 0xc82112f460, 0x1, 0x1, 0xc82112f401)
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/controller.go:294 +0x38e
github.com/docker/libnetwork.(*controller).processNodeDiscovery(0xc8208a54a0, 0xc82112f460, 0x1, 0x1, 0x7f49361e6101)
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/controller.go:277 +0x1e7
github.com/docker/libnetwork.(*controller).hostJoinCallback(0xc8208a54a0, 0xc82112f460, 0x1, 0x1)
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/controller.go:261 +0x44
github.com/docker/libnetwork.(*controller).(github.com/docker/libnetwork.hostJoinCallback)-fm(0xc82112f460, 0x1, 0x1)
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/controller.go:250 +0x3e
github.com/docker/libnetwork/hostdiscovery.(*hostDiscovery).processCallback(0xc8207c9c20, 0xc82160d0e0, 0x2, 0x2, 0xc8207fecc0, 0xc8207fecd0, 0xc8207fece0)
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/hostdiscovery/hostdiscovery.go:86 +0x19b
github.com/docker/libnetwork/hostdiscovery.(*hostDiscovery).monitorDiscovery(0xc8207c9c20, 0xc820846360, 0xc820846420, 0xc8207fecc0, 0xc8207fecd0, 0xc8207fece0)
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/hostdiscovery/hostdiscovery.go:54 +0x146
created by github.com/docker/libnetwork/hostdiscovery.(*hostDiscovery).Watch
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/hostdiscovery/hostdiscovery.go:45 +0x13f

goroutine 28 [IO wait, 84 minutes]:
net.runtime_pollWait(0x7f4936234608, 0x72, 0xc82000e110)
        /usr/local/go/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc820294450, 0x72, 0x0, 0x0)
        /usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc820294450, 0x0, 0x0)
        /usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).readMsg(0xc8202943f0, 0xc8206524a0, 0x10, 0x10, 0xc8207f0420, 0x1000, 0x1000, 0xffffffffffffffff, 0x0, 0x0, ...)
        /usr/local/go/src/net/fd_unix.go:286 +0x33d
net.(*UnixConn).ReadMsgUnix(0xc820020140, 0xc8206524a0, 0x10, 0x10, 0xc8207f0420, 0x1000, 0x1000, 0xc81ffc07df, 0x0, 0xc81ffc07df, ...)
        /usr/local/go/src/net/unixsock_posix.go:150 +0x15f
github.com/godbus/dbus.(*oobReader).Read(0xc8207f0400, 0xc8206524a0, 0x10, 0x10, 0x88, 0x0, 0x0)
        /usr/src/docker/vendor/src/github.com/godbus/dbus/transport_unix.go:21 +0xbf
io.ReadAtLeast(0x7f49361e65b8, 0xc8207f0400, 0xc8206524a0, 0x10, 0x10, 0x10, 0x0, 0x0, 0x0)
        /usr/local/go/src/io/io.go:298 +0xe6
io.ReadFull(0x7f49361e65b8, 0xc8207f0400, 0xc8206524a0, 0x10, 0x10, 0x51, 0x0, 0x0)
        /usr/local/go/src/io/io.go:316 +0x62
github.com/godbus/dbus.(*unixTransport).ReadMessage(0xc8207fee50, 0xc8207c9e60, 0x0, 0x0)
        /usr/src/docker/vendor/src/github.com/godbus/dbus/transport_unix.go:85 +0x1bc
github.com/godbus/dbus.(*Conn).inWorker(0xc8208bb320)
        /usr/src/docker/vendor/src/github.com/godbus/dbus/conn.go:248 +0x55
created by github.com/godbus/dbus.(*Conn).Auth
        /usr/src/docker/vendor/src/github.com/godbus/dbus/auth.go:118 +0xf50

goroutine 29 [chan receive, 84 minutes]:
github.com/godbus/dbus.(*Conn).outWorker(0xc8208bb320)
        /usr/src/docker/vendor/src/github.com/godbus/dbus/conn.go:385 +0x55
created by github.com/godbus/dbus.(*Conn).Auth
        /usr/src/docker/vendor/src/github.com/godbus/dbus/auth.go:119 +0xf75

goroutine 183 [chan receive]:
github.com/docker/docker/daemon.(*statsCollector).run(0xc820a6fd80)
        /usr/src/docker/.gopath/src/github.com/docker/docker/daemon/stats_collector_unix.go:99 +0xcd
created by github.com/docker/docker/daemon.(*Daemon).newStatsCollector
        /usr/src/docker/.gopath/src/github.com/docker/docker/daemon/stats_collector_unix.go:38 +0x1af

goroutine 184 [chan receive, 4 minutes]:
github.com/docker/docker/daemon.(*Daemon).execCommandGC(0xc820001c80)
        /usr/src/docker/.gopath/src/github.com/docker/docker/daemon/exec.go:237 +0x8e
created by github.com/docker/docker/daemon.NewDaemon
        /usr/src/docker/.gopath/src/github.com/docker/docker/daemon/daemon.go:807 +0x30f1

goroutine 107 [select]:
github.com/docker/libnetwork/osl.removeUnusedPaths()
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/osl/namespace_linux.go:74 +0x490
created by github.com/docker/libnetwork/osl.createBasePath
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/osl/namespace_linux.go:59 +0xbe

goroutine 144 [IO wait, 84 minutes]:
net.runtime_pollWait(0x7f4936234488, 0x72, 0xc82000e110)
        /usr/local/go/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc8201f7410, 0x72, 0x0, 0x0)
        /usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc8201f7410, 0x0, 0x0)
        /usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).accept(0xc8201f73b0, 0x0, 0x7f4936232f28, 0xc820959380)
        /usr/local/go/src/net/fd_unix.go:408 +0x27c
net.(*UnixListener).AcceptUnix(0xc820959120, 0x14, 0x0, 0x0)
        /usr/local/go/src/net/unixsock_posix.go:304 +0x53
net.(*UnixListener).Accept(0xc820959120, 0x0, 0x0, 0x0, 0x0)
        /usr/local/go/src/net/unixsock_posix.go:314 +0x41
github.com/docker/libnetwork.(*controller).acceptClientConnections(0xc8208a54a0, 0xc8201f7340, 0x63, 0x7f4936233790, 0xc820959120)
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/sandbox_externalkey_unix.go:130 +0x45
created by github.com/docker/libnetwork.(*controller).startExternalKeyListener
        /usr/src/docker/vendor/src/github.com/docker/libnetwork/sandbox_externalkey_unix.go:124 +0x28a

....

**Comment by mavenugo** [comment]
Thanks @jpetazzo @pdevine. We have a proposed fix in #1050. We will look into the root-cause and get this resolved for 1.11. thanks.
